doc/contribute/contributor_expectations.rst

```diff
@@ -87,27 +87,37 @@ Changes which require an RFC proposal include:
 Maintainers have the discretion to request that contributors create an RFC for
 PRs that are too large or complicated.
 
+.. _pr_requirements:
+
 PR Requirements
 ***************
 
 - Each commit in the PR must provide a commit message following the
   :ref:`commit-guidelines`.
 
+- No fixup or merge commits are allowed, see :ref:`Contribution workflow` for
+  more information.
+
 - The PR description must include a summary of the changes and their rationales.
 
 - All files in the PR must comply with :ref:`Licensing
   Requirements<licensing_requirements>`.
 
-- Follow the Zephyr :ref:`coding_style` and :ref:`coding_guidelines`.
+- The code must follow the Zephyr :ref:`coding_style` and :ref:`coding_guidelines`.
 
-- PRs must pass all CI checks. This is a requirement to merge the PR.
+- The PR must pass all CI checks, as described in :ref:`merge_criteria`.
   Contributors may mark a PR as draft and explicitly request reviewers to
   provide early feedback, even with failing CI checks.
 
 - When breaking a PR into multiple commits, each commit must build cleanly. The
   CI system does not enforce this policy, so it is the PR author's
   responsibility to verify.
 
+- When breaking up a PR into multiple commits, the sequence of commits must
+  follow a logical order as per the changes introduced, and not reflect the
+  order in which the changes were developed or modified. This includes not
```

Comment:
the wording currently reads as "the commits must follow a logical order" (makes sense) AND "must not reflect the order  in which the changes were developed". There is no reason I think that it must not reflect development order, that might be impractical, if the logical/reviewing order also matches the development order. I think the wording here is too strong and coupled, maybe it should say like:

"Commits in a pull request must represent distinct, reviewable, logical units of change that retain bisectibility. Unrelated changes should not be present in the same commit. Intermediary changes or non-final artifacts that occur in the git history during development of a pull request must be squashed out of these fundamental commits."

Reply:
This reads great @decsny! I will use it verbatim, thanks!.

Reply:
wow

Reply:
If Alice, Bob, and Charlie can't get along then maybe they just don't work together in the first place

Reply:
@decsny, please... Just see recent comments in https://github.com/zephyrproject-rtos/zephyr/pull/78795 as an example of the situation being described by @ajf58, and please try to refrain from this kind of innuendo. Thanks!

Reply:
> Please correct me if I'm wrong.
> 
> [...]
> Likewise, Alice might not actually want to signoff on Charlie's work. So who should be in the Signed-off-by? Is the committer going to assume Alice, Bob, and Charlie sign off on the squashed mega-commit?

I understand what you're trying to say. The sign off is not really meant as a sign off of the work/code that went in the commit, but rather as a sign off of the Developer Certificate of Origin. There is no requirement that all authors sign off, as whoever ends up signing off basically certifies that to the best of their knowledge the commit is OK license wise etc. More at https://docs.zephyrproject.org/latest/contribute/guidelines.html#developer-certification-of-origin-dco

That being said, and also pointed out in the PR I linked just above, this doesn't solve the attribution part of the problem, at least not at the commit level (files always have copyright header tracking copyright holders, so that part is definitely covered as per your comment re: Apache license). 

When commit sequence and granularity doesn't hinder bisectability and has a reasonable logic to it, I think it should be fine to let original author(s) decide to keep commits unsquashed if that helps preserving "git blame" traceability (your point about wanting a git-native solution).

Reply:
If someone takes an existing commit/PR then adds to it and it's part of the same thing, the changes should be squashed and a second signed off line added as per https://docs.zephyrproject.org/latest/contribute/guidelines.html#dco-sign-off

> Additional requirements:
>     If you are altering an existing commit created by someone else, you must add your Signed-off-by: line without removing the existing one.

There should not be multiple commits just to keep one original if it adds absolutely nothing or as per other PRs just gets completed changed in following commits as part of the same PR

Reply:
For now I will stay away from the commit attribution/ownership debate because I don't feel I will be able to do a good job combining it with what is, in my opinion, the key topic here: the subdivision of logical work units in commits.